Refactor and support new "single stage" auto click method

ronzulu opened the pull request:
This introduces "single stage" auto click functionality into clickless mouse.

The existing "two stage" functionality has not been changed, and is set as the default method.

The rationale for this new functionality is that (for me at least) 95% of the time when moving the mouse to a new spot, the intention is to left click at that point. So this is what then new functionality does by default, with a voice command to override.

This has been implemented by:

1. creating a  "clicker_base" base class for clicker functionality

1. extracting the existing two stage code into a two_stage_clicker class

1. creating a new single_stage_clicker class

There is more work I need to do on the single stage clicker class, but was wanting to show you what has being developed so far and see if it's functionality you would like to include.

Cheers
Ronny

ronzulu committed:
progress

ronzulu committed:
progress

ronzulu committed:
progress

ronzulu committed:
progress

ronzulu committed:
progress

ronzulu committed:
progress

ronzulu committed:
progress

ronzulu committed:
progress

ronzulu committed:
refactor in preparation of new functionality

ronzulu committed:
initial implementation of single_stage_clicker

ronzulu committed:
progress

ronzulu committed:
minor

ronzulu committed:
separated out mouse position analysis into new class mouse_state_analyzer

ronzulu committed:
minor

ronzulu committed:
don't auto click if the user is pressing or has just completed pressing a mouse button

knausj85 commented:
I think this functionality makes a lot of sense! I took a cursory look at the refactor too and it looks good as well; I'll take a closer look once you move it out of draft. 

Thanks for taking the time to contribute to this one!

ronzulu commented:
Great, glad to hear. The concept came from a utility called "MouseTool" which I've been using for 20+ years. I'll have to include an acknowledgement (although their website is no longer around). Having it integrated into talon and controllable by voices is great.

When I've been using the updated click less mouse in `single-stage` mode, there have been times when it has clicked on something near the location of the mouse pointer, but not something I intended.

I'll need to get to the bottom of this before finalizing the PR.

Cheers

ronzulu committed:
progress

ronzulu committed:
progress with left drag

ronzulu committed:
left drag repeat

ronzulu committed:
implemented end repeat

ronzulu committed:
progress

ronzulu committed:
minor

ronzulu committed:
comment

ronzulu committed:
restore default clicker method

ronzulu committed:
minor

ronzulu commented:
I figured out the issue regarding the spurious clicks, it wasn't a coding issue, or user error, not sure how to categorize it!

Basically I'm running talon within VM displaying within a window. So sometimes I would move the mouse outside of the vm window. When that happens, the guest os keeps reporting the last mouse position within the window. Hence the code would see that the mouse had stopped moving and performed the click.

I've added a few more features, and it's ready for review. I've added another phrase to invoke the voice commands with "clicker" in addition to a your "click less". If you don't like it, then I'll remove it and just keep that in my own personal customizations.

I haven't updated the documentation yet, hopefully do it soonish.

knausj85 commented on clickless_mouse.talon:
```diff
@@ -14,4 +31,10 @@ settings():
+^(click less | clicker) disable$: user.clickless_mouse_disable()
+
+# command specific to single_stage
+^(click less | clicker) {user.clickless_mouse_action}$: 
```
I wonder if we should have tags for when the single vs two stage clicker is active, and split this command out into a separate talon file? This way, it's only active when appropriate.

knausj85 commented on standstill_action.py:
```diff
@@ -0,0 +1,5 @@
+
+mod = Module()
+
+mod.list("clickless_mouse_action", "The list of actions that can be performed")
```
Since this is specific to the single stage clicker, I might be inclined to move this definition into its python file.

And maybe rename to like clickless_mouse_single_stage_actions? Maybe not heh, that's very verbose.

knausj85 commented on standstill_action.talon-list:
```diff
@@ -0,0 +1,28 @@
+list: user.clickless_mouse_action
```
I generally recommend naming the file the same as the list name for ease of searchability

knausj85 commented on standstill_action.talon-list:
```diff
@@ -0,0 +1,28 @@
+
+hover: hover
+
+duke: left_dubclick
```
Some of these might conflict with default community commands. I'm not sure how or if we should we worry about it. Just noting for when I get around to testing

knausj85 reviewed (changes requested):
Few minor notes, looks great. I'll need to put it thru a few paces before merging, probably next weekend.

ronzulu commented:
Thanks for all the feedback! I'll definitely look at each one in detail.

I'm still seeing times  where something unexpected happens on my PC, and I can't immediately rule out the cause being a spurious click happening.

So I'm in the process of adding detailed logging  every time a mouse click is stimulated. It might be a week or so...
